### PR TITLE
Custom config file install path

### PR DESCRIPTION
**Pull request by fbuihuu**
This is a draft because only the header of system.conf  has been updated for now. just want to be sure that these changes will be accepted before updating the rest of the headers.

### CODE REVIEW

**Comment by poettering** [comment]
hmm, didn't we have this discussion before?

**Comment by fbuihuu** [comment]
We discussed about making the installation of main config files as drop-ins previously and optionally symlinking them for backward compatibility reason.

Here it's a proposal for customizing the installation path of the main config files (without doing any convertion). We already allow to not install them, I'm not sure to understand why it would be a bad thing to allow distros to install the templates in /usr/lib/systemd instead of /etc/systemd.

The goal is to reduce the interactions of rpm with /etc when sytemd package is updated.

**Comment by arvidjaar** [comment]
> I'm not sure to understand why it would be a bad thing to allow distros to install the templates in /usr/lib/systemd instead of /etc/systemd.

a) this will give false impression that systemd is using these files while it is not. So some packagers will try to change the distribution-wide defaults by changing this file and will complain that it does not work.
b) this location is not documented so no user will even try to read them.

**Comment by Werkov** [comment]
I think `systemd-analyze cat-config` would display the modified location (thanks to macro `CONF_PATHS_NULSTR`). (But configs should be really parsed too to avoid confusion.)

Then these files (presumably RO with all lines commented out and the actual config under `/etc` via dropins) could serve users to show what the compiled in defaults are -- I don't think there's an alternative nowadays (values in `/etc` may diverge).

**Comment by fbuihuu** [comment]
@poettering I dropped the controversial commit. I don't see the reason why we shouldn't  allow downstream to customize the installation path of main config files but at least the 2 other commits should encounter less resistance.

Can you please have a look and if it's good for you I'll finish to update the headers of the main config files.

Thanks.

**Review by keszybz — APPROVED** [review]
LGTM.

**Comment by keszybz** [comment]
I merged the two documentation patches. As yo said, those changes are not controversial.

As for the third patch: I agree that we should move our config out of `/etc/`, or at least that we should make the path configurable. The idea of having defaults in `/etc/systemd/something.conf` was good for its time, but that was before we had drop-ins. Nowadays, the recommended way for local configuration is **not** to modify that file, but to create drop-ins. Effectively this means that the only effective way of viewing configuration is with `systemd-analyze cat-config`, which also means that the file with commented-out settings might just as well be under `/usr`. For viewing this makes no difference. But it would allow us to remove stuff from `/etc`. It would also discourage users from modifying the main config file, which is an added benefit.

Please submit the last patch again, so we can discuss things.

**Comment by keszybz** [comment]
> only the header of system.conf has been updated for now

Ah, I missed that. Anyway, please update the other ones too.

**Comment by fbuihuu** [comment]
@keszybz thanks for your feedback. I'll resubmit the last patch.

**Comment by fbuihuu** [comment]
> Ah, I missed that. Anyway, please update the other ones too.

I was about to pointing out that the changes you merged weren't finished. That's the reason why I set this PR as "draft" to somehow prevent it from being merged.

I'll submit another commit to finish the work soon.

**Review by YHNdnzj — CHANGES_REQUESTED** [review]
Please make paths to directories end with `/`

**Comment by fbuihuu** [comment]
> Please make paths to directories end with `/`

You meant all of them ? For example /etc/ or /usr/ look a bit weird but maybe it's just me.

**Comment by fbuihuu** [comment]
> Please submit the last patch again, so we can discuss things.

https://github.com/systemd/systemd/pull/29244

**Comment by poettering** [comment]
This should not have been merged. I think this is just wrong.

File: man/standard-conf.xml [hunk 1]
```diff
@@ -42,11 +42,12 @@
     <title>Configuration Directories and Precedence</title>
 
     <para>The default configuration is set during compilation, so configuration is only needed when it is
-    necessary to deviate from those defaults. Initially, the main configuration file in
-    <filename>/etc/systemd/</filename> contains commented out entries showing the defaults as a guide to the
-    administrator. Local overrides can be created by editing this file or by creating drop-ins, as described
-    below. Using drop-ins for local configuration is recommended over modifications to the main configuration
-    file.</para>
+    necessary to deviate from those defaults. The main configuration file is either in
+    <filename>/usr/lib/systemd</filename> or <filename>/etc/systemd/</filename> and contains commented out
```
Comment (poettering) [inline man/standard-conf.xml:46]: where is the "main configuration file in /usr/lib/systemd"? that's made up, just wrong.

Comment (poettering) [reply man/standard-conf.xml:46]: this does not use our usual rule that dir paths should be suffixed by /

File: man/standard-conf.xml [hunk 2]
```diff
@@ -42,11 +42,12 @@
...
+    entries showing the defaults as a guide to the administrator. Local overrides can be created by creating
+    drop-ins, as described below. The main configuration file can also be edited for this purpose (or a copy
+    in <filename>/etc</filename> if it's shipped in <filename>/usr</filename>) however using drop-ins for
```
Comment (poettering) [inline man/standard-conf.xml:49]: here too, slash is missing.

also, we do not ship "main config files" in /usr/, so it doesn't make sense to copy them over.

**Comment by poettering** [comment]
also, generally: if you change entirely what a PR does, rename it. THe PR title does not match what was merged at all. This is all very confusing.

**Comment by poettering** [comment]
> As for the third patch: I agree that we should move our config out of `/etc/`, or at least that we should make the path configurable. 

Nope. We have no config in `/etc/`. Just some helpful hints for people who want a populated fully populated `/etc/`.

These files should not be shipped in environments that have an empty /etc/.

> The idea of having defaults in `/etc/systemd/something.conf` was good for its time, but that was before we had drop-ins. Nowadays, the recommended way for local configuration is **not** to modify that file, but to create drop-ins. 

That's simply not true. Why would they do that? editing the main config file is totally OK. 

Sorry, but this is nonsensical. And apparently no agreement on all of this.

**Comment by keszybz** [comment]
> That's simply not true. Why would they do that? editing the main config file is totally OK.

It is certainly supported and it works. But it is not the best approach. It is much better to use a drop-in. This is particularly true for stuff like `systemd.conf`: the user almost always wants to override one or two options, so a drop-in is much more suitable.

> also, we do not ship "main config files" in /usr/, so it doesn't make sense to copy them over.

We support loading config files from `/usr` and `/usr/local`. As written, I think the text is not wrong. It is careful to say that the file may be in one place or the other and that if it is in the other place, in needs to be copied before editing.

**Comment by fbuihuu** [comment]
> That's simply not true. Why would they do that? editing the main config file is totally OK.

No, it's not OK because in this case it prevents packages from relying on drop-ins (to customize upstream defaults). This problem has been reported several time already. And the only way out is to rely on drop-ins exclusively.

**Comment by poettering** [comment]
I am sorry, but we shouldn't tell people the "main configuration file is in /usr/lib/". That's just a totally nonsensical thing to say. Maybe on suse you configure stuff in /usr/lib now, but that's really not how this is intended to be used from our upstream perspective. We put configuration files in /etc/, and they are redundant by default, which means on stateless systems the clean choice is to simply not have them by default. 

Sorry, this really should be reverted.

**Comment by fbuihuu** [comment]
> I am sorry, but we shouldn't tell people the "main configuration file is in /usr/lib/". That's just a totally nonsensical thing to say. Maybe on suse you configure stuff in /usr/lib now, but that's really not how this is intended to be used from our upstream perspective. 

I don't see where the idea of "SUSE configures main config files in /usr/lib" is coming from.

Again we want to move them in /usr/lib simply to make them available as *templates* for users and also to make `systemd-analyze cat-config` still continue to show the defaults names/values without any modification in the tool.

If `systemd-analyze cat-config` showed these defaults, we wouldn't need to ship the templates in /usr/lib.

**Comment by fbuihuu** [comment]
@poettering @keszybz could we please take a decision ?

If shipping the main config in /usr as *templates* is not something that going to happen, can we at least find a solution to make the templates available somehow when the installation of the main config files is disabled ?

As already suggested, the defaults could be optionally part of the output of `systemd-analyze cat-config` for example.

Thanks.

**Comment by keszybz** [comment]
I tried to lay out my thinking in https://github.com/systemd/systemd/pull/29244#pullrequestreview-1637180262. I still think that your patch is the best way to approach this: the only new functionality is a new config switch which allows distros to move files to a different place, there is nothing new to learn for users, and the new workflow is nice. In particular, it works nicely with the approach used in Fedora and other places where the same package build (with some set of defaults) is used to construct a bunch of different variants (Workstation, Server, IoT, containers, etc.), and a drop-in mechanism is used to override the defaults, and those drop-ins should be located in `/usr/` because they are part of a static payload of packages. 

Adding a new mechanism to "remember" and show the defaults (`--show-defaults` or so) is possible, but I don't find it attractive. In particular, if `fedora-workstation-release.rpm` adds a drop-in which overrides the built-in default in `systemd.rpm`, what should be shown? In the approach where we show all config files in precedence order, we don't even need to answer this question: you explicitly see both the commented out the default and the variant override. And the nice thing is that it's all using the same generic mechanism, so it's self-explanatory to anyone who grokked how drop-ins work.

The third option is *status quo*. I think we can do better, but some people disagree. The nice thing is that with the your patch people can continue to do the old thing.

**Comment by fbuihuu** [comment]
> Adding a new mechanism to "remember" and show the defaults (`--show-defaults` or so) is possible, but I don't find it attractive. In particular, if `fedora-workstation-release.rpm` adds a drop-in which overrides the built-in default in `systemd.rpm`, what should be shown? 

The same as before: the built-in defaults (all entries commented out) followed by the drop-ins (in precedence order). I don't think we should treat the drop-ins shipped by downstream specifically.

Actually a more useful option might be needed to shorten the output of `systemd-anlaze` by skipping all commented entries. That would allow one to see the actual overrides more easily.

> In the approach where we show all config files in precedence order, we don't even need to answer this question: you explicitly see both the commented out the default and the variant override. And the nice thing is that it's all using the same generic mechanism, so it's self-explanatory to anyone who grokked how drop-ins work.

The main reason for suggesting to treat the built-in defaults specifically was to find an alternative solution due to the fact that @poettering doesn't like the idea to install the templates in /usr.

**Comment by keszybz** [comment]
@fbuihuu Can you submit a patch that adds the same advice blob to other files?

**Pull request merged**